task/attest/src/main.rs

```diff
@@ -0,0 +1,185 @@
+                [0..self.cert_data.persistid_cert.0.size as usize]),
+            3 => {
+                if let Some(cert) = self.cert_data.intermediate_cert.as_ref() {
+                    Ok(&cert.0.as_bytes()[0..cert.0.size as usize])
+                } else {
+                    Err(AttestError::InvalidCertIndex.into())
+                }
+            }
+            _ => Err(AttestError::InvalidCertIndex.into()),
+        }
+    }
+}
+
+impl idl::InOrderAttestImpl for AttestServer<'_> {
+    /// Get length of cert chain from Alias to mfg intermediate
+    fn cert_chain_len(
+        &mut self,
+        _: &userlib::RecvMessage,
+    ) -> Result<u32, RequestError<AttestError>> {
+        // The cert chain will vary in length:
+        // - kernel w/ feature 'dice-self' will have 3 certs in the chain w/
+        // the final cert being a self signed, puf derived identity key
+        // - kernel /w feature 'dice-mfg' will have 4 certs in the chain w/
+        // the final cert being the intermediate that signs the identity
+        // cert
+        let chain_len = if self.cert_data.intermediate_cert.is_none() {
+            3
+        } else {
+            4
+        };
+
+        ringbuf_entry!(Trace::CertChainLen(chain_len));
+        Ok(chain_len)
+    }
+
+    /// Get length of cert at provided index in cert chain
+    fn cert_len(
+        &mut self,
+        _: &userlib::RecvMessage,
+        index: u32,
+    ) -> Result<usize, RequestError<AttestError>> {
+        let len = self.get_cert_bytes_from_index(index)?.len();
+
+        ringbuf_entry!(Trace::CertLen(len));
+        Ok(len)
+    }
+
+    /// Get a cert from the AliasCert chain
+    fn cert(
+        &mut self,
+        _: &userlib::RecvMessage,
+        index: u32,
+        offset: usize,
+        dest: Leased<W, [u8]>,
+    ) -> Result<(), RequestError<AttestError>> {
+        ringbuf_entry!(Trace::Cert);
+        ringbuf_entry!(Trace::Index(index));
+        ringbuf_entry!(Trace::Offset(offset));
+        ringbuf_entry!(Trace::BufSize(dest.len()));
+
+        let cert = self.get_cert_bytes_from_index(index)?;
+        if cert.is_empty() {
+            let err = AttestError::InvalidCertIndex;
+            ringbuf_entry!(Trace::Error(err));
+            return Err(err.into());
+        }
+
+        // there must be sufficient data read from cert to fill the lease
+        if dest.len() > cert.len() - offset {
+            let err = AttestError::OutOfRange;
+            ringbuf_entry!(Trace::Error(err));
+            return Err(err.into());
+        }
+
+        dest.write_range(0..dest.len(), &cert[offset..offset + dest.len()])
+            .map_err(|_| RequestError::Fail(ClientError::WentAway))?;
+
+        Ok(())
+    }
+}
+
+#[export_name = "main"]
+fn main() -> ! {
+    ringbuf_entry!(Trace::Startup);
+
+    let addr = unsafe { CERTS.assume_init_ref() };
+    let cert_data = match CertData::load_from_addr(addr) {
+        Ok(a) => a,
+        Err(_) => panic!("CertData"),
```

Comment:
If the cert data cannot be loaded, this will cause your task to restart forever, starving all tasks at lower priorities. That's not good.

I recommend instead changing the content of the `Server` to `Option` and providing a way of indicating "could not load certs" in the IPC API.

Reply:
oof that's bad test coverage on my part.

Reply:
ran across a small bug in `lib-lpc55-rot-startup` dependencies while testing this: https://github.com/oxidecomputer/hubris/pull/1402